### ISSUE TITLE
mishandling of git-annex external special remote protocol unsupported request

### ISSUE THREAD

**Issue by joeyh** [datalad/datalad#2124]
To use new special remote protocol messages like the one added in
http://git-annex.branchable.com/todo/INFO_message_for_custom_special_remotes/ , external special remotes need a way to determine if git-annex supports the command. http://git-annex.branchable.com/design/external_special_remote_protocol has a plan to make git-annex send a new message after VERSION indicating its capabilities.

On auditing all external special remotes to see if they will work well with such a new message, I found all did, except for the one in datalad, which does this:

            method = getattr(self, "req_%s" % req, None)
            if not method:
                self.error("We have no support for %s request, part of %s response"
                           % (req, l))
                self.send("UNSUPPORTED-REQUEST")
                continue

As far as I can tell, this probably crashes. It should simply send back the UNSUPPORTED-REQUEST and let git-annex deal with it.

**Comment by yarikoptic** [comment]
Thanks @joeyh for heads up.  I am not sure why should it crash -- if annex sends some command which we have no support (`req_` method), then `getattr` returns None and we send `UNSUPPORTED-REQUEST` to annex, so we should be ok.

But I am still wondering why not to maintain compatibility by relying on special remotes requesting specific version/features support, like I have outlined in http://git-annex.branchable.com/design/exporting_trees_to_special_remotes/#comment-b3876cc7b791c6b8cb12949bdbbf8268

That would preclude braking compatibility with any existing external remote which might not have resilience to unsupported requests, while allowing for needed flexibility

**Comment by joeyh** [comment]
Doesn't the self.error make it crash?

**Comment by yarikoptic** [comment]
I don't think so -- just reports to our log and sends ERROR message to parent git-annex.
Also it would make no sense then to `self.send` in the next line, would it? ;)

**Comment by joeyh** [comment]
Well, if it sends ERROR to git-annex, git-annex will then stop talking to it entirely, which is just as bad as the backtrace I was assuming.

**Comment by yarikoptic** [comment]
oh... indeed.I should analyze our use of `self.error` then -- probably I should come up with a `.warning` which would send an `INFO` to annex with `WARNING: `-prefixed message, unless you see it worthwhile adding `WARNING` into special remote protocol.
thanks for the note

**Comment by joeyh** [comment]
git-annex deals with UNSUPPORTED-REQUEST on its own so you should not need to send back any INFO or do anything except send that.

**Comment by joeyh** [comment]
Check out git-annex's new EXTENSIONS branch to see what happens when git-annex sends some unsupported request..

**Comment by yarikoptic** [comment]
d'oh -- apparently I have dealt with this one  awhile back in 9d58db6b67525b75ae1a79677a6fbf5e811a49a8 which is (0.9.1~5^2~3) and thus current code is not like in original quote (I guess taken prior 0.9.1):
```python
            method = getattr(self, "req_%s" % req, None)
            if not method:
                self.debug("We have no support for %s request, part of %s response"
                           % (req, l))
                self.send("UNSUPPORTED-REQUEST")
                continue
```
so we are Ok'ish.  But there are other places where we use `self.error` where we might need to reconsider.  I will close this issue pushing some minor tune ups on that end

**Comment by joeyh** [comment]
Ah ok, and current versions of git-annex send EXPORTSUPPORTED which triggers the same problem code in old datalad. But, that's done only if the datalad remote is configured for export, which I think it will never be, so that incompatability avoiding breaking it.

Remains the question of what to do about the old versions of datalad that would be broken if I merge the EXTENSIONS branch into git-annex. Could make the git-annex package conflict with them or something..

**Comment by yarikoptic** [comment]
IMHO that would be ok to demand datalad >=0.9.1 ;-). At least on neuroDebians we assure to have datalad compatible with available through apt git annex

**Comment by yarikoptic** [comment]
was fixed in #2125
